File: src/dlgs/wnd_picture_funcs.rs
use winsafe::{self as w, prelude::*};

use super::WndPicture;
use crate::{id3v2, msgbox};

impl WndPicture {
	pub(super) fn load_picture(&self, sel_tags: &[id3v2::Tag]) -> w::AnyResult<()> {
		if !id3v2::equal_frame_across_all_tags("APIC", sel_tags) {
			return Ok(());
		}

		match &sel_tags[0].frame_by_name4("APIC").unwrap().body() {
			id3v2::Body::Picture(pic_body) => {
				let hparent = self.wnd.hwnd().GetParent()?;

				let stream = match w::SHCreateMemStream(&pic_body.data) {
					Err(err) => {
						msgbox::err(
							&hparent,
							"Cannot load picture stream",
							None,
							&format!("SHCreateMemStream error:\n{}", &err.to_string()),
						)?;
						self.pic_err.set(Some(err)); // store the error
						return Ok(());
					},
					Ok(stream) => stream,
				};

				let pic_obj = match w::OleLoadPicture(&stream, None, true) {
					Err(err) => {
						msgbox::err(
							&hparent,
							"Cannot parse picture",
							None,
							&format!("OleLoadPicture error:\n{}", &err.to_string()),
						)?;
						self.pic_err.set(Some(err)); // store the error
						return Ok(());
					},
					Ok(pic_obj) => pic_obj,
				};

				*self.pic.try_borrow_mut()? = Some(pic_obj);
				self.wnd.hwnd().InvalidateRect(None, true)?;
			},
			_ => return Err("APIC body is not picture.".into()), // should never happen
		}

		Ok(())
	}

	pub(super) fn unload_picture(&self) -> w::AnyResult<()> {
		*self.pic.try_borrow_mut()? = None;
		self.wnd.hwnd().InvalidateRect(None, true)?;
		Ok(())
	}
}

File: wndpicture/WndPictureFuncs.go
//go:build windows

package wndpicture

import (
	"id3fit/id3v2"

	"github.com/rodrigocfd/windigo/ui"
	"github.com/rodrigocfd/windigo/win"
)

// Loads the cover art, if due, into the IPicture COM object.
func (me *WndPicture) LoadPicture(tags []*id3v2.Tag) (pixels win.SIZE, nBytes int) {
	apic := id3v2.SameFrameAcrossAllTags("APIC", tags)
	if apic == nil {
		return win.SIZE{}, 0 // we don't have a picture to display
	}

	body, ok := apic.Body().(*id3v2.BodyPicture)
	if !ok {
		ui.MsgError(me.wnd.Parent(), "Picture parsing", "",
			"APIC frame does not contain BodyPicture body type") // should never happen
		return win.SIZE{}, 0
	}

	localOleRel := win.NewOleReleaser()
	defer localOleRel.Release()

	iStream, err := win.SHCreateMemStream(localOleRel, body.Bin) // create IStream over pic data
	if err != nil {
		ui.MsgError(me.wnd.Parent(), "Picture stream", "",
			"Failed to stream picture:\n"+err.Error())
		return win.SIZE{}, 0
	}

	me.oleRel.ReleaseNow(me.iPic) // free IPicture right away, before loading new

	me.iPic, err = win.OleLoadPicture(me.oleRel, iStream, len(body.Bin), true)
	if err != nil {
		ui.MsgError(me.wnd.Parent(), "Picture loading", "",
			"Failed to load picture:\n"+err.Error())
		return win.SIZE{}, 0
	}

	hdcScreen, _ := win.HWND(0).GetDC()
	defer win.HWND(0).ReleaseDC(hdcScreen)
	szPic, _ := me.iPic.SizePixels(hdcScreen) // picture resolution in pixels

	return szPic, len(body.Bin)
}
